SU2_CFD/include/solvers/CScalarSolver.hpp

```diff
@@ -0,0 +1,294 @@
+
+#ifdef HAVE_OMP
+  vector<GridColor<> > EdgeColoring; /*!< \brief Edge colors. */
+  bool ReducerStrategy = false;      /*!< \brief If the reducer strategy is in use. */
+#else
+  array<DummyGridColor<>, 1> EdgeColoring;
+  /*--- Never use the reducer strategy if compiling for MPI-only. ---*/
+  static constexpr bool ReducerStrategy = false;
+#endif
+
+  /*--- Edge fluxes for reducer strategy (see the notes in CEulerSolver.hpp). ---*/
+  CSysVector<su2double> EdgeFluxes; /*!< \brief Flux across each edge. */
+
+  /*!
+   * \brief The highest level in the variable hierarchy this solver can safely use.
+   */
+  CScalarVariable* nodes = nullptr;
+
+  /*!
+   * \brief Return nodes to allow CSolver::base_nodes to be set.
+   */
+  inline CVariable* GetBaseClassPointerToNodes() final { return nodes; }
+
+  /*!
+   * \brief Compute the viscous flux for the turbulent equation at a particular edge.
+   * \tparam SolverSpecificNumericsFunc - Struct, that implements solver specific contributions to numerics.
+   * \note The functor has to implement (*geometry, *numerics, *config, *nodes, iPoint, jPoint)
+   * \param[in] iEdge - Edge for which we want to compute the flux
+   * \param[in] geometry - Geometrical definition of the problem.
+   * \param[in] solver_container - Container vector with all the solutions.
+   * \param[in] numerics - Description of the numerical method.
+   * \param[in] config - Definition of the particular problem.
+   */
+  template <class SolverSpecificNumericsFunc>
+  FORCEINLINE void Viscous_Residual_impl(const SolverSpecificNumericsFunc& SolverSpecificNumerics, unsigned long iEdge,
+                                         CGeometry* geometry, CSolver** solver_container, CNumerics* numerics,
+                                         CConfig* config) {
+    const bool implicit = (config->GetKind_TimeIntScheme() == EULER_IMPLICIT);
+    CVariable* flowNodes = solver_container[FLOW_SOL]->GetNodes();
+
+    /*--- Points in edge ---*/
+
+    auto iPoint = geometry->edges->GetNode(iEdge, 0);
+    auto jPoint = geometry->edges->GetNode(iEdge, 1);
+
+    /*--- Points coordinates, and normal vector ---*/
+
+    numerics->SetCoord(geometry->nodes->GetCoord(iPoint), geometry->nodes->GetCoord(jPoint));
+    numerics->SetNormal(geometry->edges->GetNormal(iEdge));
+
+    /*--- Conservative variables w/o reconstruction ---*/
+
+    numerics->SetPrimitive(flowNodes->GetPrimitive(iPoint), flowNodes->GetPrimitive(jPoint));
+
+    /*--- Turbulent variables w/o reconstruction, and its gradients ---*/
+
+    numerics->SetScalarVar(nodes->GetSolution(iPoint), nodes->GetSolution(jPoint));
+    numerics->SetScalarVarGradient(nodes->GetGradient(iPoint), nodes->GetGradient(jPoint));
+
+    /*--- Call Numerics contribution which are Solver-Specifc. Implemented in the caller: Viscous_Residual.  ---*/
+
+    SolverSpecificNumerics(*geometry, *numerics, *config, *nodes, iPoint, jPoint);
```

Comment:
I would have passed just iPoint / jPoint because the caller has everything already.